cpu/native/ng_tapnet/ng_tapnet.c

```diff
@@ -0,0 +1,703 @@
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @ingroup     cpu_native_ng_tapnet
+ * @{
+ *
+ * @file
+ *
+ * @author      Ludwig Ortmann <ludwig.ortmann@fu-berlin.de>
+ * @author      Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#ifdef __MACH__
+#define _POSIX_C_SOURCE
+#include <net/if.h>
+#undef _POSIX_C_SOURCE
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+
+#elif defined(__FreeBSD__)
+#include <sys/socket.h>
+#include <net/if.h>
+#include <ifaddrs.h>
+#include <net/if_dl.h>
+
+#else
+#include <net/if.h>
+#include <linux/if_tun.h>
+#include <linux/if_ether.h>
+#endif
+
+#include "byteorder.h"
+#include "native_internal.h"
+#include "net/ng_ethernet.h"
+#include "net/ng_ethertype.h"
+#include "net/ng_netdev.h"
+#include "net/ng_netif/hdr.h"
+#include "net/ng_pkt.h"
+#include "net/ng_pktbuf.h"
+#include "od.h"
+#include "thread.h"
+#include "utlist.h"
+
+#include "ng_tapnet.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _ISR_EVENT_RX   (1U)
+
+ng_tapnet_t ng_tapnet;
+
+static uint8_t send_buffer[NG_ETHERNET_MAX_LEN];
+static uint8_t recv_buffer[NG_ETHERNET_MAX_LEN];
+
+/* MAC OSX specific variables and functions */
+#ifdef __MACH__
+pid_t _sigio_child_pid;
+
+static void _sigio_child(ng_tapnet_t *dev);
+#endif
+
+/* driver function definitions */
+static int _send_data(ng_netdev_t *netdev, ng_pktsnip_t *pkt);
+static int _add_event_callback(ng_netdev_t *dev, ng_netdev_event_cb_t cb);
+static int _rem_event_callback(ng_netdev_t *dev, ng_netdev_event_cb_t cb);
+static int _get(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t max_len);
+static int _set(ng_netdev_t *dev, ng_netconf_opt_t opt, void *value,
+                size_t value_len);
+static void _isr_event(ng_netdev_t *dev, uint32_t event_type);
+
+/* netdev driver struct */
+const ng_netdev_driver_t ng_tapnet_driver = {
+    _send_data,
+    _add_event_callback,
+    _rem_event_callback,
+    _get,
+    _set,
+    _isr_event,
+};
+
+/* internal function definitions */
+static inline bool _is_addr_broadcast(uint8_t *addr)
+{
+    return ((addr[0] == 0xff) && (addr[1] == 0xff) && (addr[2] == 0xff) &&
+            (addr[3] == 0xff) && (addr[4] == 0xff) && (addr[5] == 0xff));
+}
+
+static inline bool _is_addr_multicast(uint8_t *addr)
+{
+    /* source: http://ieee802.org/secmail/pdfocSP2xXA6d.pdf */
+    return (addr[0] & 0x01);
+}
+
+/* build Ethernet packet from pkt */
+static int _marshall_ethernet(ng_tapnet_t *dev, uint8_t *buffer, ng_pktsnip_t *pkt);
+
+/* build ISR handler for ISR events */
+void _trigger_isr_event(void);
+
+/* driver implementation */
```

Comment:
Then you might want to make this a more obvious file section marker. I used something like this in some parts of native:
```
/**********************************************
 * foobar
 **********************************************/